File: Templates/KMP.java

class KMPTemplate{

    static int[] constructLPS(String pat){
        int m = pat.length();
        int[] lps = new int[m];
        int i = 0;
        int j = 1;
        while (j < m){
            if(pat.charAt(i) == pat.charAt(j)){
                lps[j] = i + 1;
                i++;
                j++;
            }
            else{
                if(i != 0){
                    i = lps[i - 1];
                }
                else{
                    lps[j] = 0;
                    j++;
                }
            }
        }
        return lps;
    }

    static boolean KMP(String a, String pat){
        int[] lps = constructLPS(pat);
        int i = 0;
        int j = 0;
        while (j < pat.length() && i < a.length()){
            if(a.charAt(i) == pat.charAt(j)){
                j++;
                i++;
            }
            else{
                if(j != 0) {
                    j = lps[j - 1];
                }
                else{
                    i++;
                }
            }
        }
        return j == pat.length();
    }
}

File: Templates/KMP.py
from typing import List

# KMP Algorithm 
def constructLPS(pat: str) -> List[int]:
    m = len(pat)
    lps = [0] * m
    i = 0
    j = 1
    while j < m:
        if pat[i] == pat[j]:
            lps[j] = i + 1
            i += 1
            j += 1
        else:
            if i != 0:
                i = lps[i - 1]
            else:
                lps[j] = 0
                j += 1
    return lps


def KMP(a: str, pat: str):
    lps = constructLPS(pat)
    i = 0
    j = 0
    while j < len(pat) and i < len(a):
        if a[i] == pat[j]:
            j += 1
            i += 1
        else:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1
    return j == len(pat)

